### PR TITLE
cks: fix template, deployment issues

### LINKED ISSUE #4056: The listKubernetesSupportedVersions API does not honour searching
The kubernetes ISOs or the listKubernetesSupportedVersions API does not support searching as the keyword parameter is not honoured. Fixing this would enable searching both in UI and API.

Originally opened as Primate bug - https://github.com/apache/cloudstack-primate/issues/280

##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
API, UI
~~~

##### CLOUDSTACK VERSION

~~~
4.14
~~~

### LINKED ISSUE #4107: CKS: Same name (previously deleted template) template for node VMs fails k8s cluster deployment
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
CKS
~~~

##### CLOUDSTACK VERSION

~~~
4.14.0
~~~

##### SUMMARY

When CKS is set to use a CoreOS template with a name which is same as an already deleted template, plugin fails to deploy Kubernetes clusters using new, active template

##### STEPS TO REPRODUCE

~~~
1. Enable CKS plugin
2. Register CoreOS template
3. Delete that template after installation
4. Add the template again with the same name
5. Change the cloud.kubernetes.cluster.template.name.*  with the template name
6. Verify template is successfully installed
7. Register Kubernetes version ISO and try to deploy a k8s cluster
~~~

##### EXPECTED RESULTS

~~~
Cluster should be deployed successfully using recently registered, active CoreOS template
~~~

##### ACTUAL RESULTS

~~~
Cluster deployment fails,
2020-05-27 09:03:55,080 DEBUG [c.c.a.ApiServlet] (qtp1104443373-22:ctx-610114b3) (logid:3f96511a) ===START===  10.1.0.1 -- GET  command=createKubernetesCluster&response=json&name=K8s-cluster1&description=K8s-cluster1&zoneid=6c099729-4848-4165-bb30-53764af7a245&kubernetesversionid=fc6953b6-c1e7-49f1-81be-2661c804e059&serviceofferingid=e377c887-b5e9-41e6-851b-ea7f37115ee3&size=3&keypair=&masternodes=1&_=1590568220199
2020-05-27 09:03:55,084 DEBUG [c.c.a.ApiServer] (qtp1104443373-22:ctx-610114b3 ctx-294f0797) (logid:3f96511a) CIDRs from which account 'Acct[18042dac-9c0b-11ea-8607-1e004f01080f-admin]' is allowed to perform API calls: 0.0.0.0/0,::/0
2020-05-27 09:03:55,090 WARN  [c.c.k.c.KubernetesClusterManagerImpl] (qtp1104443373-22:ctx-610114b3 ctx-294f0797) (logid:3f96511a) The template ID: 199c5101-cbbb-481a-8319-7d06329d7461, name: coreos-kvm is not available for use in zone ID: 6c099729-4848-4165-bb30-53764af7a245 provisioning Kubernetes cluster nodes
2020-05-27 09:03:55,090 INFO  [c.c.a.ApiServer] (qtp1104443373-22:ctx-610114b3 ctx-294f0797) (logid:3f96511a) Kubernetes service has not been configured properly to provision Kubernetes clusters
2020-05-27 09:03:55,090 DEBUG [c.c.a.ApiServlet] (qtp1104443373-22:ctx-610114b3 ctx-294f0797) (logid:3f96511a) ===END===  10.1.0.1 -- GET  command=createKubernetesCluster&response=json&name=K8s-cluster1&description=K8s-cluster1&zoneid=6c099729-4848-4165-bb30-53764af7a245&kubernetesversionid=fc6953b6-c1e7-49f1-81be-2661c804e059&serviceofferingid=e377c887-b5e9-41e6-851b-ea7f37115ee3&size=3&keypair=&masternodes=1&_=1590568220199
~~~
image (1)

Related code:
https://github.com/apache/cloudstack/blob/master/plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java#L298
https://github.com/apache/cloudstack/blob/master/engine/schema/src/main/java/com/cloud/storage/dao/VMTemplateDaoImpl.java#L133-L137

### LINKED ISSUE #4113: CKS: Multi-hypervisor zone doesn't adhere offering tags and fails deployment
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
CKS
~~~

##### CLOUDSTACK VERSION

~~~
4.14.0
~~~

##### CONFIGURATION

Advance networking

##### OS / ENVIRONMENT

N/A

##### SUMMARY

With a zone having clusters of different hypervisor hosts it is not possible to select hypervisor in which k8s cluster to be deployed.
If offering tags are set such that to deploy k8s cluster in specific hypervisor then as well it tries to deploy k8s cluster in the undesirable hypervisor and eventually fails with either template error or during cluster node VM deployment.

##### STEPS TO REPRODUCE

~~~

- Setup an environment with the Advanced zone having different hypervisor hosts/clusters. Eg, KVM and Xenserver. Xen cluster should be added in the DB first.
- Enable CKS, install CoreOS template for KVM and register k8s version
- Add a compute offering with host tags to use KVM hosts
- Try to deploy a k8s cluster with new offering

~~~

##### EXPECTED RESULTS

~~~
k8s cluster gets deployed successfully
~~~

##### ACTUAL RESULTS

~~~
Error: Kubernetes service has not been configured properly to provision Kubernetes clusters

2020-05-26 19:56:46,006 WARN  [c.c.k.c.KubernetesClusterManagerImpl] (qtp1104443373-22:ctx-fd12355a ctx-5faee980) (logid:48a3af28) The template ID: 199c5101-cbbb-481a-8319-7d06329d7461, name: coreos-kvm is not available for use in zone ID: 6c099729-4848-4165-bb30-53764af7a245 provisioning Kubernetes cluster nodes
2020-05-26 19:56:46,006 INFO  [c.c.a.ApiServer] (qtp1104443373-22:ctx-fd12355a ctx-5faee980) (logid:48a3af28) Kubernetes service has not been configured properly to provision Kubernetes clusters
2020-05-26 19:56:46,007 DEBUG [c.c.a.ApiServlet] (qtp1104443373-22:ctx-fd12355a ctx-5faee980) (logid:48a3af28) ===END===  10.1.0.1 -- GET  command=createKubernetesCluster&response=json&name=k8s-cl1&description=k8s-cl1&zoneid=6c099729-4848-4165-bb30-53764af7a245&kubernetesversionid=fc6953b6-c1e7-49f1-81be-2661c804e059&serviceofferingid=e377c887-b5e9-41e6-851b-ea7f37115ee3&size=1&keypair=&noderootdisksize=10&masternodes=1&_=1590523030807

If CoreOS template is also added for XenServer before deploying k8s cluster then plugin fails while deploying node VMs for cluster due to incompatible offering and hosts.

246242 2020-05-26 20:39:58,557 DEBUG [c.c.d.FirstFitPlanner] (API-Job-Executor-45:ctx-49f7e115 job-414 ctx-a370d486) (logid:c2c543bf) No clusters found after removing disabled clusters and clusters in av       oid list, returning.
246243 2020-05-26 20:39:58,560 DEBUG [c.c.a.ApiServlet] (qtp1104443373-20:ctx-08683735) (logid:be313ec6) ===START===  10.1.0.1 -- GET  command=queryAsyncJobResult&jobId=c2c543bf-e7fb-4552-9f05-28ade5ff8e       c1&response=json&_=1590523031173
246244 2020-05-26 20:39:58,563 ERROR [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-45:ctx-49f7e115 job-414 ctx-a370d486) (logid:c2c543bf) Provisioning the master VM failed in the Kubernetes        cluster ID: 0f6922f6-116d-40cd-894b-01d63eed8ead
246245 com.cloud.exception.ManagementServerException: Failed to start VM in the Kubernetes cluster ID: 0f6922f6-116d-40cd-894b-01d63eed8ead                                                                246246 >---at com.cloud.kubernetes.cluster.actionworkers.KubernetesClusterResourceModifierActionWorker.startKubernetesVM(KubernetesClusterResourceModifierActionWorker.java:271)
~~~

### LINKED ISSUE #4133: CKS: Cluster remains in destroying state when non-cluster VMs are using cluster network
##### ISSUE TYPE

 * Improvement Request

##### COMPONENT NAME

~~~
CKS
~~~

##### CLOUDSTACK VERSION

~~~
4.14.0
~~~

##### OS / ENVIRONMENT

N/A

##### SUMMARY

When additional VMs are added to the k8s cluster network, cluster deletion fails with a message, Unable to delete cluster network. k8s cluster remains in destroying state. It gets removed once non-cluster VMs are migrated from the cluster network. Instead of failing during deletion itself, service can intimate user before deletion that the network is being used by other VMs and they must be migrated before deleting.

##### STEPS TO REPRODUCE

~~~
1. Create a k8s cluster without passing an existing network.
2. Create and add a new VM to this network outside of CKS
3. Delete k8s cluster
~~~

##### EXPECTED RESULTS

~~~
Service should throw an error before deleting k8s cluster that cluster network is in use by VMs outside k8s cluster. So these VMs can be migrated.
~~~

##### ACTUAL RESULTS

~~~
Service tries to delete the k8s cluster. Node VMs are deleted. But cannot delete the network. Error is thrown at this time. Cluster shows up in destroying state until VMs are migrated from the network.
~~~

### PR DESCRIPTION
**Pull request by shwstppr**
## Description
Fixes #4056 
Fixes #4107
Fixes #4113
Fixes #4133 

Check for compute offering host tags while finding suitable host/deployment destination for k8s cluster node VM during deploy and scale cluster.
Check for k8s cluster template's hypervisor type during scale as the template is tied to the k8s cluster therefore new node VM must be deployed in the same hypervisor host.

Fixes template checks

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

File: engine/schema/src/main/java/com/cloud/storage/dao/VMTemplateDao.java [hunk 1]
```diff
@@ -36,6 +36,8 @@
 
     public VMTemplateVO findByTemplateName(String templateName);
 
+    public VMTemplateVO findValidByTemplateName(String templateName);
+
     // public void update(VMTemplateVO template);
```
Comment (rhtyd) [inline engine/schema/src/main/java/com/cloud/storage/dao/VMTemplateDao.java:41]: Remove unused/commented method

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java [hunk 2]
```diff
@@ -908,6 +910,10 @@ private void validateKubernetesClusterUpgradeParameters(UpgradeKubernetesCluster
         if (!KubernetesCluster.State.Running.equals(kubernetesCluster.getState())) {
             throw new InvalidParameterValueException(String.format("Kubernetes cluster ID: %s is not in running state", kubernetesCluster.getUuid()));
         }
+        final DataCenter zone = dataCenterDao.findById(kubernetesCluster.getZoneId());
+        if (zone == null) {
+            logAndThrow(Level.WARN, String.format("Unable to find zone for Kubernetes cluster ID: %s", kubernetesCluster.getUuid()));
```
Comment (rhtyd) [inline plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java:915]: Is `logAndThrow` a new utility?

Comment (shwstppr) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java:915]: Just some helper methods to log and throw exceptions for similar error https://github.com/apache/cloudstack/blob/8c6861bdc0f6b9fbf58a1c099f36ba08348dd204/plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java#L216-L249

File: engine/schema/src/main/java/com/cloud/storage/dao/VMTemplateDao.java (hunk 1 above)
Comment (shwstppr) [reply engine/schema/src/main/java/com/cloud/storage/dao/VMTemplateDao.java]: done

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java [hunk 3]
```diff
@@ -300,8 +294,7 @@ private boolean isKubernetesServiceTemplateConfigured(DataCenter zone) {
                 LOGGER.warn(String.format("Unable to find the template %s to be used for provisioning Kubernetes cluster nodes", templateName));
                 return false;
             }
-            List<VMTemplateZoneVO> listZoneTemplate = templateZoneDao.listByZoneTemplate(zone.getId(), template.getId());
-            if (listZoneTemplate == null || listZoneTemplate.isEmpty()) {
+            if (CollectionUtils.isEmpty(templateJoinDao.newTemplateView(template, zone.getId(), true))) {
```
Comment (DaanHoogland) [inline plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java:297]: The call `public List<TemplateJoinVO> newTemplateView(VirtualMachineTemplate template, long zoneId, boolean readyOnly)` as implemented in `TemplateJoinDaoImpl`, does a search including removed leading to a hit for a template that is not available. Is this intended?

**Comment by rhtyd** [comment]
@shwstppr can you club your cks PRs, and at least change base branch to 4.14?

**Comment by shwstppr** [comment]
@blueorangutan package

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java (hunk 3 above)
Comment (shwstppr) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java:300]: @DaanHoogland thanks for pointing it. I'll check that again

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
Can you help review and test @davidjumani

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@shwstppr is this ready for review, can you open is for review then?
@davidjumani @Pearl1594 @DaanHoogland can you help review as well

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java (hunk 3 above)
Comment (DaanHoogland) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java:300]: any resolution on this @shwstppr ?

Comment (shwstppr) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java:300]: @DaanHoogland I've checked this. `newTemplateView` is called here with `readyOnly=true` and deleted templates have `state=NULL`.

**Pull request merged**